bee-bundle/src/bundle/outgoing_bundle_builder.rs

```diff
@@ -58,17 +74,79 @@ where
     S: OutgoingBundleBuilderStage,
 {
     // TODO TEST
-    fn calculate_hash(&self) -> TritBuf {
-        // TODO Impl
+    fn calculate_bundle_hash(&mut self) -> Result<(), OutgoingBundleBuilderError> {
         let mut sponge = E::default();
+        let mut obsolete_tag = match self.builders.0.get(0) {
+            Some(builder) => match builder.obsolete_tag.as_ref() {
+                Some(obsolete_tag) => obsolete_tag.to_inner().to_owned(),
+                _ => {
+                    return Err(OutgoingBundleBuilderError::MissingTransactionBuilderField(
+                        "obsolete_tag",
+                    ))
+                }
+            },
+            _ => return Err(OutgoingBundleBuilderError::Empty),
+        };
+
+        let hash = loop {
+            sponge.reset();
+
+            for builder in &self.builders.0 {
+                let _ = sponge.absorb(&builder.essence());
+            }
+
+            let hash = sponge
+                .squeeze()
+                .unwrap_or_else(|_| panic!("Panicked when unwrapping the sponge hash function."));
+
+            let hash = normalize_hash(&hash);
+            let mut has_m_bug = false;
+            for trits in hash.chunks(3) {
+                let mut is_m = true;
+
+                for trit in trits.iter() {
+                    if *trit != Btrit::PlusOne {
+                        is_m = false;
+                        break;
+                    }
+                }
+
+                if is_m {
+                    has_m_bug = true;
+                    break;
+                }
+            }
 
-        for _builder in &self.builders.0 {
-            // TODO sponge.absorb(builder.essence());
+            if !has_m_bug {
+                break Hash::from_inner_unchecked(hash);
+            } else {
+                // obsolete_tag + 1
+                for i in 0..obsolete_tag.len() {
```

Comment:
Please add a comment that we may want to move that addition thing to the ternary crate